### DISCUSSION TITLE
sdpvar and 'for loop'

### DISCUSSION THREAD

**Discussion by HuliSaer** [yalmip/YALMIP#1126] [General discussion] [answered]
**Dear Johan Löfberg,
It was a pleasure to be able to reach you by email and thank you very much for providing the Yalmip toolbox for helping me solve a lot of problems.
But I've had some trouble lately about the relationship between variables and loops in Yalmip.
First, the operation process is to segment the output of each generator. Second, each segment has a carbon emission coefficient, which is used to calculate the carbon emissions of each segment. Finally, the cumulative sum of carbon emissions for all segments is calculated to find the minimum value.
The code is as follows:**
`clear all`
`gennum=1;%the number of generator`
`T=2;%time`
`p=sdpvar(gennum,T,'full');%the output of generator`
`%%`
`P_1=0;%Initial value`
`p_1=0.5;%Interval length`
`PL=[5 6];%load`
`%%`
`st=[];`
`for t=1:T`
`st=[st,sum(p(:,t))==PL(1,t)];
%The generator generates the same amount as the user's electricity consumption`
`end`
`for t=1:T`
    `for i=1:gennum`
    `n(i,t)=floor((p(i,t)-P_1)./p_1)+1;%sublevel`  
   ` for q=1:n(i,t)`
      ` E_q(t,q)=max(0.06-(q-1)*0.01,0);
%Calculate the carbon emission coefficient for each segment`
      `end`
     `E_c(i,t)=sum(E_q(t,:))`
     `end`
`end`
`totalcost=0`
`for t=1:T`
`for i=1:gennum`
`  totalcost=totalcost+E_c(i,t);`
`end`
`end`
`result=optimize(st,totalcost);`
**However, the result of the operation will be reported in line 17:
The result is that the operator ':' corresponding to the input parameter of type 'sdpvar' is not defined.
If I modify 17 lines of code：**
`for q=1:value（n(i,t)）;`
**The result of the run is that the index must be of positive integer type or logical type.
Both of the above methods will show`q=NAN`;
I know that the optimal function needs to calculate the carbon emission coefficient of each segment based on the number of segments n, in this case the type of n is sdpvar.
Why is the value of q different from n?
I would like to ask how to solve this problem, and I would appreciate it if you could help me.Thank you very much！**

**Comment by johanlofberg** [comment, marked as answer]
You will have to express the stuff without using indexing which depends on decision variables.

Your code looks very convoluted. simplify the expression to something easily expressed using logical expressions and if-statements

y = x(1:t) where both x and t are decision variables, would for instance be written as

if t = 1 then y = x(1)
if t = 2 then y = sum(x(1:2))

etc

which then is written using MILP-representable stuff

https://yalmip.github.io/modellingif/

However, your code looks overly complex and most likely you are modelling something very simple piecewise affine where you don't need floor and indexing etc, you basically just have something like

case 1: p is in some region, E has some value
case 2: p is in some other , E has some other value
etc

which once again is modelled using same method as above, just logical implications

**Comment by HuliSaer** [comment]
> You will have to express the stuff without using indexing which depends on decision variables.
> 
> Your code looks very convoluted. simplify the expression to something easily expressed using logical expressions and if-statements
> 
> y = x(1:t) where both x and t are decision variables, would for instance be written as
> 
> if t = 1 then y = x(1) if t = 2 then y = sum(x(1:2))
> 
> etc
> 
> which then is written using MILP-representable stuff
> 
> https://yalmip.github.io/modellingif/
> 
> However, your code looks overly complex and most likely you are modelling something very simple piecewise affine where you don't need floor and indexing etc, you basically just have something like
> 
> case 1: p is in some region, E has some value case 2: p is in some other , E has some other value etc
> 
> which once again is modelled using same method as above, just logical implications

Thank you soooooooooooooo much！Thank you for your help！！！！！！

**Comment by johanlofberg** [comment]
Your function is just a piecewise constant function which thus can be used when setting up a model using implies over the regions

```
P_1=0;%Initial value
p_1=0.5;%Interval length
E_c = [];
for p = 0:0.01:15
    n=floor((p-P_1)./p_1)+1;
    E_q  = 0;
    for q=1:n
        E_q = E_q + max(0.06-(q-1)*0.01,0); %Calculate the carbon emission coefficient for each segment
    end
    E_c=[E_c E_q];
end
stairs(0:0.01:15,E_c)
```

**Comment by hyy660** [comment]
> > You will have to express the stuff without using indexing which depends on decision variables.
> > Your code looks very convoluted. simplify the expression to something easily expressed using logical expressions and if-statements
> > y = x(1:t) where both x and t are decision variables, would for instance be written as
> > if t = 1 then y = x(1) if t = 2 then y = sum(x(1:2))
> > etc
> > which then is written using MILP-representable stuff
> > https://yalmip.github.io/modellingif/
> > However, your code looks overly complex and most likely you are modelling something very simple piecewise affine where you don't need floor and indexing etc, you basically just have something like
> > case 1: p is in some region, E has some value case 2: p is in some other , E has some other value etc
> > which once again is modelled using same method as above, just logical implications
> 
> Thank you soooooooooooooo much！Thank you for your help！！！！！！



> > You will have to express the stuff without using indexing which depends on decision variables.
> > Your code looks very convoluted. simplify the expression to something easily expressed using logical expressions and if-statements
> > y = x(1:t) where both x and t are decision variables, would for instance be written as
> > if t = 1 then y = x(1) if t = 2 then y = sum(x(1:2))
> > etc
> > which then is written using MILP-representable stuff
> > https://yalmip.github.io/modellingif/
> > However, your code looks overly complex and most likely you are modelling something very simple piecewise affine where you don't need floor and indexing etc, you basically just have something like
> > case 1: p is in some region, E has some value case 2: p is in some other , E has some other value etc
> > which once again is modelled using same method as above, just logical implications
> 
> Thank you soooooooooooooo much！Thank you for your help！！！！！！
hello friend, i have also encountered the problem of indexing sdpvar, have you solved it?
